replace macro script with autoclass methods

carshadi opened the pull request:
- better to autoclass in the relevant java classes rather than throw an entire script into a macro.
- added visualization of the convex hull with Viewer3D

carshadi committed:
replace macro script with autoclass methods

- better to autoclass in the relevant java classes rather than throw an entire script into a macro.
- added visualization of the convex hull with Viewer3D

ctrueden commented on src/main/resources/script_templates/Neuroanatomy/pyimagej/pyijtest.py:
```diff
@@ -1,67 +1,82 @@
+    verts = X[hull.vertices]
+
+    # construct new java Hashset containing the hull vertices
+    verts_set = HashSet()
```
You can write `ij.py.to_java(set(verts))` to go from Python set to Java set.

carshadi replied:
Hi @ctrueden thank you for the tips. I suppose the difficulty lies in converting between interfaces specific to Fiji, in this case SNTPoint, and python data types. From my limited perusing, I wasn't able to find a way to do this within scyjava. If this is already supported please let me know.

A couple alternative options come to mind:

option 1.
```
tips_java_set = analyzer.getTips()
tips_java_array = tips_java_set.toArray()
```
then use ``` hull.vertices ``` to index the specific vertices from ``` tips_java_array ```.  I am not sure whether accessing java arrays by index bypass the calls over JNI, but it would forgo having to construct a new PointInImage() Java object for each computed vertex.

option 2:
add a method within SNT to accept a collection of ArrayLists and convert into a SNTPoint set. 

@tferr thoughts?

carshadi replied:
option 3: implement the SNTPoint interface from within python using jnius.PythonJavaClass

carshadi replied:
would option 3 provide a performance gain over using autoclass(tracing.util.PointInImage)?

ctrueden reviewed (commented):
I'm not sure I agree that it's better to code in Python with fine-grained calls to Java API. Every call to Java is a call over JNI which has potential performance implications, especially if done in big loops. It is also possible that certain Java API calls may crash the program (i.e. segfault).

In the end, the most important thing is that it is tested and working with reasonable performance. If so, probably fine to ignore my warnings above.

tferr commented:
>  It is also possible that certain Java API calls may crash the program (i.e. segfault).

Nah. [pyimagej](https://github.com/imagej/pyimagej) is so incredible, that will never happen :) This would be a template script as a proof-of-principle that it is possible to analyze SNT data in numpy, a request we get often. It does not impact on SNT functionality directly. @carshadi's effort was also useful to shape SNT's API to be more script-friendly.

But now that I think about it, having it as a Script Editor's Template may not be useful as it is not a jython script. @ctrueden, do you think we should place it somewhere else?

tferr commented on src/main/resources/script_templates/Neuroanatomy/pyimagej/pyijtest.py:
```diff
@@ -1,67 +1,82 @@
-import ast
 from scipy.spatial import ConvexHull
 
-ij = imagej.init(r'C:\Users\cam\Desktop\Fiji.app', headless=False)
```
It would be useful to mention that this can be ` ij = imagej.init('sc.fiji:fiji')` once the scijava branch gets merged into master

carshadi replied:
done

ctrueden commented:
> pyimagej is so incredible, that will never happen

I'm happy you like pyimagej. :-) But speaking from experience, it is not hard to crash the Python kernel with cells as simple as:
```python
from jnius import autoclass
IJ = autoclass('ij.IJ')
IJ.getImage()
```
(at least for me, on macOS)

> do you think we should place it somewhere else?

I agree that putting it in `script_templates` is misleading, since it will not run as-is from the Script Editor. So putting it somewhere else makes sense to me.

tferr commented:
Hi @carshadi ,

The performance profiling was great! At this point, I would propose the following:
1. Have a _performance_testing_ notebook with your findings (for those who want/need to debug and optimize code)
2. Have a simpler notebook that makes the convex hull calculations (for those that are just interested in morphometry analysis)

As per discussion above, both could be moved to a _notebooks_ folder at the root of the project. Let me know if you need help or feedback

carshadi commented:
sure can do

carshadi committed:
create notebooks for performance testing and convex hull

- create one jupyter notebook to quantify performance of converting between Java and Python data types
- create another jupyter notebook to perform the convex hull calculation and rendering
- removed the initial commit which is less performant 

The convex hull of a neuron reconstruction is one way of quantifying how much space a cell takes up within a compartment(s). 
A future addition could be to compute the convex hulls of 2+ cells and measure the amount of overlap,
in order to assess spatial groupings of terminal node clusters.
